diff --git a/Cargo.lock b/Cargo.lock
index 5db3c6d55..a54c97cce 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -3024,9 +3024,9 @@ dependencies = [
 
 [[package]]
 name = "psm"
-version = "0.1.16"
+version = "0.1.21"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "cd136ff4382c4753fc061cb9e4712ab2af263376b95bbd5bd8cd50c020b78e69"
+checksum = "5787f7cda34e3033a72192c018bc5883100330f362ef279a8cbccfce8bb4e874"
 dependencies = [
  "cc",
 ]
@@ -4998,9 +4998,9 @@ checksum = "a8f112729512f8e442d81f95a8a7ddf2b7c6b8a1a6f509a95864142b30cab2d3"
 
 [[package]]
 name = "stacker"
-version = "0.1.14"
+version = "0.1.15"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "90939d5171a4420b3ff5fbc8954d641e7377335454c259dcb80786f3f21dc9b4"
+checksum = "c886bd4480155fd3ef527d45e9ac8dd7118a898a46530b7b94c3e21866259fce"
 dependencies = [
  "cc",
  "cfg-if 1.0.0",
diff --git a/compiler/rustc_llvm/build.rs b/compiler/rustc_llvm/build.rs
index 7729ec6be..ffbf90cf3 100644
--- a/compiler/rustc_llvm/build.rs
+++ b/compiler/rustc_llvm/build.rs
@@ -137,7 +137,6 @@ fn main() {
         "aarch64",
         "amdgpu",
         "avr",
-        "m68k",
         "mips",
         "powerpc",
         "systemz",
diff --git a/compiler/rustc_llvm/src/lib.rs b/compiler/rustc_llvm/src/lib.rs
index 8eade02a4..ad3cb421f 100644
--- a/compiler/rustc_llvm/src/lib.rs
+++ b/compiler/rustc_llvm/src/lib.rs
@@ -92,14 +92,6 @@ pub fn initialize_available_targets() {
         LLVMInitializeAVRAsmPrinter,
         LLVMInitializeAVRAsmParser
     );
-    init_target!(
-        llvm_component = "m68k",
-        LLVMInitializeM68kTargetInfo,
-        LLVMInitializeM68kTarget,
-        LLVMInitializeM68kTargetMC,
-        LLVMInitializeM68kAsmPrinter,
-        LLVMInitializeM68kAsmParser
-    );
     init_target!(
         llvm_component = "mips",
         LLVMInitializeMipsTargetInfo,
diff --git a/library/panic_unwind/src/gcc.rs b/library/panic_unwind/src/gcc.rs
index a0297b4b2..8db74f66b 100644
--- a/library/panic_unwind/src/gcc.rs
+++ b/library/panic_unwind/src/gcc.rs
@@ -263,7 +263,7 @@ cfg_if::cfg_if! {
         }
 
         cfg_if::cfg_if! {
-            if #[cfg(all(windows, target_arch = "x86_64", target_env = "gnu"))] {
+            if #[cfg(all(windows, any(target_arch = "aarch64", target_arch = "x86_64"), target_env = "gnu"))] {
                 // On x86_64 MinGW targets, the unwinding mechanism is SEH however the unwind
                 // handler data (aka LSDA) uses GCC-compatible encoding.
                 #[lang = "eh_personality"]
diff --git a/library/unwind/src/libunwind.rs b/library/unwind/src/libunwind.rs
index 7b78bda42..8eb4a4484 100644
--- a/library/unwind/src/libunwind.rs
+++ b/library/unwind/src/libunwind.rs
@@ -37,7 +37,7 @@ pub const unwinder_private_data_size: usize = 20;
 pub const unwinder_private_data_size: usize = 5;
 
 #[cfg(all(target_arch = "aarch64", target_pointer_width = "64"))]
-pub const unwinder_private_data_size: usize = 2;
+pub const unwinder_private_data_size: usize = 6;
 
 #[cfg(all(target_arch = "aarch64", target_pointer_width = "32"))]
 pub const unwinder_private_data_size: usize = 5;
@@ -90,7 +90,13 @@ pub type _Unwind_Exception_Cleanup_Fn =
 // rustc_codegen_ssa::src::back::symbol_export, rustc_middle::middle::exported_symbols
 // and RFC 2841
 #[cfg_attr(
-    all(feature = "llvm-libunwind", any(target_os = "fuchsia", target_os = "linux")),
+    any(
+        all(
+            feature = "llvm-libunwind",
+            any(target_os = "fuchsia", target_os = "linux")
+        ),
+        all(target_os = "windows", target_env = "gnu", target_abi = "llvm")
+    ),
     link(name = "unwind", kind = "static", modifiers = "-bundle")
 )]
 extern "C-unwind" {
@@ -269,7 +275,7 @@ if #[cfg(not(all(target_os = "ios", target_arch = "arm")))] {
 } // cfg_if!
 
 cfg_if::cfg_if! {
-if #[cfg(all(windows, target_arch = "x86_64", target_env = "gnu"))] {
+if #[cfg(all(windows, any(target_arch = "aarch64", target_arch = "x86_64"), target_env = "gnu"))] {
     // We declare these as opaque types. This is fine since you just need to
     // pass them to _GCC_specific_handler and forget about them.
     pub enum EXCEPTION_RECORD {}
diff --git a/src/bootstrap/dist.rs b/src/bootstrap/dist.rs
index b1fae356d..053844791 100644
--- a/src/bootstrap/dist.rs
+++ b/src/bootstrap/dist.rs
@@ -1495,10 +1495,6 @@ impl Step for Extended {
             };
             prepare("rustc");
             prepare("cargo");
-            prepare("rust-analysis");
-            prepare("rust-docs");
-            prepare("rust-std");
-            prepare("clippy");
             for tool in &["rust-demangler", "rls", "rust-analyzer", "miri"] {
                 if built_tools.contains(tool) {
                     prepare(tool);
@@ -1507,349 +1503,10 @@ impl Step for Extended {
             if target.ends_with("windows-gnu") {
                 prepare("rust-mingw");
             }
-
-            builder.install(&etc.join("gfx/rust-logo.ico"), &exe, 0o644);
-
-            // Generate msi installer
-            let wix_path = env::var_os("WIX")
-                .expect("`WIX` environment variable must be set for generating MSI installer(s).");
-            let wix = PathBuf::from(wix_path);
-            let heat = wix.join("bin/heat.exe");
-            let candle = wix.join("bin/candle.exe");
-            let light = wix.join("bin/light.exe");
-
-            let heat_flags = ["-nologo", "-gg", "-sfrag", "-srd", "-sreg"];
-            builder.run(
-                Command::new(&heat)
-                    .current_dir(&exe)
-                    .arg("dir")
-                    .arg("rustc")
-                    .args(&heat_flags)
-                    .arg("-cg")
-                    .arg("RustcGroup")
-                    .arg("-dr")
-                    .arg("Rustc")
-                    .arg("-var")
-                    .arg("var.RustcDir")
-                    .arg("-out")
-                    .arg(exe.join("RustcGroup.wxs")),
-            );
-            builder.run(
-                Command::new(&heat)
-                    .current_dir(&exe)
-                    .arg("dir")
-                    .arg("rust-docs")
-                    .args(&heat_flags)
-                    .arg("-cg")
-                    .arg("DocsGroup")
-                    .arg("-dr")
-                    .arg("Docs")
-                    .arg("-var")
-                    .arg("var.DocsDir")
-                    .arg("-out")
-                    .arg(exe.join("DocsGroup.wxs"))
-                    .arg("-t")
-                    .arg(etc.join("msi/squash-components.xsl")),
-            );
-            builder.run(
-                Command::new(&heat)
-                    .current_dir(&exe)
-                    .arg("dir")
-                    .arg("cargo")
-                    .args(&heat_flags)
-                    .arg("-cg")
-                    .arg("CargoGroup")
-                    .arg("-dr")
-                    .arg("Cargo")
-                    .arg("-var")
-                    .arg("var.CargoDir")
-                    .arg("-out")
-                    .arg(exe.join("CargoGroup.wxs"))
-                    .arg("-t")
-                    .arg(etc.join("msi/remove-duplicates.xsl")),
-            );
-            builder.run(
-                Command::new(&heat)
-                    .current_dir(&exe)
-                    .arg("dir")
-                    .arg("rust-std")
-                    .args(&heat_flags)
-                    .arg("-cg")
-                    .arg("StdGroup")
-                    .arg("-dr")
-                    .arg("Std")
-                    .arg("-var")
-                    .arg("var.StdDir")
-                    .arg("-out")
-                    .arg(exe.join("StdGroup.wxs")),
-            );
-            if built_tools.contains("rls") {
-                builder.run(
-                    Command::new(&heat)
-                        .current_dir(&exe)
-                        .arg("dir")
-                        .arg("rls")
-                        .args(&heat_flags)
-                        .arg("-cg")
-                        .arg("RlsGroup")
-                        .arg("-dr")
-                        .arg("Rls")
-                        .arg("-var")
-                        .arg("var.RlsDir")
-                        .arg("-out")
-                        .arg(exe.join("RlsGroup.wxs"))
-                        .arg("-t")
-                        .arg(etc.join("msi/remove-duplicates.xsl")),
-                );
-            }
-            if built_tools.contains("rust-analyzer") {
-                builder.run(
-                    Command::new(&heat)
-                        .current_dir(&exe)
-                        .arg("dir")
-                        .arg("rust-analyzer")
-                        .args(&heat_flags)
-                        .arg("-cg")
-                        .arg("RustAnalyzerGroup")
-                        .arg("-dr")
-                        .arg("RustAnalyzer")
-                        .arg("-var")
-                        .arg("var.RustAnalyzerDir")
-                        .arg("-out")
-                        .arg(exe.join("RustAnalyzerGroup.wxs"))
-                        .arg("-t")
-                        .arg(etc.join("msi/remove-duplicates.xsl")),
-                );
-            }
-            builder.run(
-                Command::new(&heat)
-                    .current_dir(&exe)
-                    .arg("dir")
-                    .arg("clippy")
-                    .args(&heat_flags)
-                    .arg("-cg")
-                    .arg("ClippyGroup")
-                    .arg("-dr")
-                    .arg("Clippy")
-                    .arg("-var")
-                    .arg("var.ClippyDir")
-                    .arg("-out")
-                    .arg(exe.join("ClippyGroup.wxs"))
-                    .arg("-t")
-                    .arg(etc.join("msi/remove-duplicates.xsl")),
-            );
-            if built_tools.contains("rust-demangler") {
-                builder.run(
-                    Command::new(&heat)
-                        .current_dir(&exe)
-                        .arg("dir")
-                        .arg("rust-demangler")
-                        .args(&heat_flags)
-                        .arg("-cg")
-                        .arg("RustDemanglerGroup")
-                        .arg("-dr")
-                        .arg("RustDemangler")
-                        .arg("-var")
-                        .arg("var.RustDemanglerDir")
-                        .arg("-out")
-                        .arg(exe.join("RustDemanglerGroup.wxs"))
-                        .arg("-t")
-                        .arg(etc.join("msi/remove-duplicates.xsl")),
-                );
-            }
-            if built_tools.contains("miri") {
-                builder.run(
-                    Command::new(&heat)
-                        .current_dir(&exe)
-                        .arg("dir")
-                        .arg("miri")
-                        .args(&heat_flags)
-                        .arg("-cg")
-                        .arg("MiriGroup")
-                        .arg("-dr")
-                        .arg("Miri")
-                        .arg("-var")
-                        .arg("var.MiriDir")
-                        .arg("-out")
-                        .arg(exe.join("MiriGroup.wxs"))
-                        .arg("-t")
-                        .arg(etc.join("msi/remove-duplicates.xsl")),
-                );
-            }
-            builder.run(
-                Command::new(&heat)
-                    .current_dir(&exe)
-                    .arg("dir")
-                    .arg("rust-analysis")
-                    .args(&heat_flags)
-                    .arg("-cg")
-                    .arg("AnalysisGroup")
-                    .arg("-dr")
-                    .arg("Analysis")
-                    .arg("-var")
-                    .arg("var.AnalysisDir")
-                    .arg("-out")
-                    .arg(exe.join("AnalysisGroup.wxs"))
-                    .arg("-t")
-                    .arg(etc.join("msi/remove-duplicates.xsl")),
-            );
-            if target.ends_with("windows-gnu") {
-                builder.run(
-                    Command::new(&heat)
-                        .current_dir(&exe)
-                        .arg("dir")
-                        .arg("rust-mingw")
-                        .args(&heat_flags)
-                        .arg("-cg")
-                        .arg("GccGroup")
-                        .arg("-dr")
-                        .arg("Gcc")
-                        .arg("-var")
-                        .arg("var.GccDir")
-                        .arg("-out")
-                        .arg(exe.join("GccGroup.wxs")),
-                );
-            }
-
-            let candle = |input: &Path| {
-                let output = exe.join(input.file_stem().unwrap()).with_extension("wixobj");
-                let arch = if target.contains("x86_64") { "x64" } else { "x86" };
-                let mut cmd = Command::new(&candle);
-                cmd.current_dir(&exe)
-                    .arg("-nologo")
-                    .arg("-dRustcDir=rustc")
-                    .arg("-dDocsDir=rust-docs")
-                    .arg("-dCargoDir=cargo")
-                    .arg("-dStdDir=rust-std")
-                    .arg("-dAnalysisDir=rust-analysis")
-                    .arg("-dClippyDir=clippy")
-                    .arg("-arch")
-                    .arg(&arch)
-                    .arg("-out")
-                    .arg(&output)
-                    .arg(&input);
-                add_env(builder, &mut cmd, target);
-
-                if built_tools.contains("rust-demangler") {
-                    cmd.arg("-dRustDemanglerDir=rust-demangler");
-                }
-                if built_tools.contains("rls") {
-                    cmd.arg("-dRlsDir=rls");
-                }
-                if built_tools.contains("rust-analyzer") {
-                    cmd.arg("-dRustAnalyzerDir=rust-analyzer");
-                }
-                if built_tools.contains("miri") {
-                    cmd.arg("-dMiriDir=miri");
-                }
-                if target.ends_with("windows-gnu") {
-                    cmd.arg("-dGccDir=rust-mingw");
-                }
-                builder.run(&mut cmd);
-            };
-            candle(&xform(&etc.join("msi/rust.wxs")));
-            candle(&etc.join("msi/ui.wxs"));
-            candle(&etc.join("msi/rustwelcomedlg.wxs"));
-            candle("RustcGroup.wxs".as_ref());
-            candle("DocsGroup.wxs".as_ref());
-            candle("CargoGroup.wxs".as_ref());
-            candle("StdGroup.wxs".as_ref());
-            candle("ClippyGroup.wxs".as_ref());
-            if built_tools.contains("rust-demangler") {
-                candle("RustDemanglerGroup.wxs".as_ref());
-            }
-            if built_tools.contains("rls") {
-                candle("RlsGroup.wxs".as_ref());
-            }
-            if built_tools.contains("rust-analyzer") {
-                candle("RustAnalyzerGroup.wxs".as_ref());
-            }
-            if built_tools.contains("miri") {
-                candle("MiriGroup.wxs".as_ref());
-            }
-            candle("AnalysisGroup.wxs".as_ref());
-
-            if target.ends_with("windows-gnu") {
-                candle("GccGroup.wxs".as_ref());
-            }
-
-            builder.create(&exe.join("LICENSE.rtf"), &rtf);
-            builder.install(&etc.join("gfx/banner.bmp"), &exe, 0o644);
-            builder.install(&etc.join("gfx/dialogbg.bmp"), &exe, 0o644);
-
-            builder.info(&format!("building `msi` installer with {:?}", light));
-            let filename = format!("{}-{}.msi", pkgname(builder, "rust"), target.triple);
-            let mut cmd = Command::new(&light);
-            cmd.arg("-nologo")
-                .arg("-ext")
-                .arg("WixUIExtension")
-                .arg("-ext")
-                .arg("WixUtilExtension")
-                .arg("-out")
-                .arg(exe.join(&filename))
-                .arg("rust.wixobj")
-                .arg("ui.wixobj")
-                .arg("rustwelcomedlg.wixobj")
-                .arg("RustcGroup.wixobj")
-                .arg("DocsGroup.wixobj")
-                .arg("CargoGroup.wixobj")
-                .arg("StdGroup.wixobj")
-                .arg("AnalysisGroup.wixobj")
-                .arg("ClippyGroup.wixobj")
-                .current_dir(&exe);
-
-            if built_tools.contains("rls") {
-                cmd.arg("RlsGroup.wixobj");
-            }
-            if built_tools.contains("rust-analyzer") {
-                cmd.arg("RustAnalyzerGroup.wixobj");
-            }
-            if built_tools.contains("rust-demangler") {
-                cmd.arg("RustDemanglerGroup.wixobj");
-            }
-            if built_tools.contains("miri") {
-                cmd.arg("MiriGroup.wixobj");
-            }
-
-            if target.ends_with("windows-gnu") {
-                cmd.arg("GccGroup.wixobj");
-            }
-            // ICE57 wrongly complains about the shortcuts
-            cmd.arg("-sice:ICE57");
-
-            let _time = timeit(builder);
-            builder.run(&mut cmd);
-
-            if !builder.config.dry_run {
-                t!(fs::rename(exe.join(&filename), distdir(builder).join(&filename)));
-            }
         }
     }
 }
 
-fn add_env(builder: &Builder<'_>, cmd: &mut Command, target: TargetSelection) {
-    let mut parts = builder.version.split('.');
-    cmd.env("CFG_RELEASE_INFO", builder.rust_version())
-        .env("CFG_RELEASE_NUM", &builder.version)
-        .env("CFG_RELEASE", builder.rust_release())
-        .env("CFG_VER_MAJOR", parts.next().unwrap())
-        .env("CFG_VER_MINOR", parts.next().unwrap())
-        .env("CFG_VER_PATCH", parts.next().unwrap())
-        .env("CFG_VER_BUILD", "0") // just needed to build
-        .env("CFG_PACKAGE_VERS", builder.rust_package_vers())
-        .env("CFG_PACKAGE_NAME", pkgname(builder, "rust"))
-        .env("CFG_BUILD", target.triple)
-        .env("CFG_CHANNEL", &builder.config.channel);
-
-    if target.contains("windows-gnullvm") {
-        cmd.env("CFG_MINGW", "1").env("CFG_ABI", "LLVM");
-    } else if target.contains("windows-gnu") {
-        cmd.env("CFG_MINGW", "1").env("CFG_ABI", "GNU");
-    } else {
-        cmd.env("CFG_MINGW", "0").env("CFG_ABI", "MSVC");
-    }
-}
-
 /// Maybe add LLVM object files to the given destination lib-dir. Allows either static or dynamic linking.
 ///
 /// Returns whether the files were actually copied.
